### PR TITLE
Fix `Parser::look_ahead`

### PR DESCRIPTION
**Pull request by nnethercote**
`Parser::look_ahead` has a slow but simple general case, and a fast special case that is hit most of the time. But the special case is buggy and behaves differently to the general case. There are also no unit tests. This PR fixes all of this, resulting in a `Parser::look_ahead` that is equally fast, slightly simpler, more correct, and better tested.

r? @davidtwco

### CODE REVIEW

**Comment by nnethercote** [comment]
Local results showed that the new code has the same performance as the old code. Let's double check that.

@bors try @rust-timer queue

File: compiler/rustc_parse/src/parser/tests.rs [hunk 1]
```diff
@@ -1376,6 +1376,127 @@ fn ttdelim_span() {
     });
 }
 
+// Uses a macro rather than a function so that failure messages mention the
+// correct line in the test function.
```
Comment (fmease) [inline compiler/rustc_parse/src/parser/tests.rs:1380]: What about `#[track_caller]` + fn?

Comment (nnethercote) [reply compiler/rustc_parse/src/parser/tests.rs:1380]: I tried that, it didn't work :/

Comment (fmease) [reply compiler/rustc_parse/src/parser/tests.rs:1380]: Ah, right. Requires `#[track_caller]` on closures (requires features `stmt_expr_attributes`, `closure_track_caller`, #87417)

File: compiler/rustc_parse/src/parser/mod.rs [hunk 2]
```diff
@@ -1118,6 +1118,35 @@ impl<'a> Parser<'a> {
...
+                None => {
+                    // The tree cursor lookahead went (one) past the end of the
+                    // current token tree. Try to return a close delimiter.
+                    if let Some(&(_, span, _, delim)) = self.token_cursor.stack.last()
+                        && delim != Delimiter::Invisible
+                    {
+                        // We are not in the outermost token stream, so we have
+                        // delimiters. Also, those delimiters are not skipped.
+                        return looker(&Token::new(token::CloseDelim(delim), span.close));
+                    }
```
Comment (oli-obk) [inline compiler/rustc_parse/src/parser/mod.rs:1145]: is this actually a common enough case for a special casing to make sense here?

File: compiler/rustc_parse/src/parser/tests.rs (hunk 1 above)
Comment (nnethercote) [reply compiler/rustc_parse/src/parser/tests.rs:1380]: Closure? I tried making `look` a top-level function like this:
```rust
#[track_caller]
fn look(p: &Parser<'_>, dist: usize, kind: rustc_ast::token::TokenKind) {
    p.look_ahead(dist, |tok| assert_eq!(kind, tok.kind));
}
```
and that's what didn't work. The assertion failure still reported the error on the line containing the `look_ahead` call. I'm not sure why it didn't work, it seems like an ideal use for `#[track_caller]`.

File: compiler/rustc_parse/src/parser/mod.rs (hunk 2 above)
Comment (nnethercote) [reply compiler/rustc_parse/src/parser/mod.rs:1145]: Yes. As per the second commit message, there are icount regressions across lots of benchmarks, up to a 2% in the worst case.

File: compiler/rustc_parse/src/parser/tests.rs (hunk 1 above)
Comment (fmease) [reply compiler/rustc_parse/src/parser/tests.rs:1380]: I'm referring to

```rs
#![feature(closure_track_caller)]

#[track_caller]
fn look(p: &Parser<'_>, dist: usize, kind: TokenKind) {
    p.look_ahead(dist, #[track_caller] |tok| assert_eq!(kind, tok.kind));
}
```

but that also points inside `look_ahead` for some reason.

**Comment by oli-obk** [comment]
r? @oli-obk

bors r plus

**Pull request merged**